xdsl/irdl/declarative_assembly_format.py

```diff
@@ -292,10 +292,11 @@ class VariadicLikeFormatDirective(
     Such directives can not be followed by comma literals.
     """
 
-    def set_empty(self, state: ParsingState):
+    def set_empty(self, state: ParsingState) -> str | None:
         """
         Set the appropriate field of the parsing state to be empty.
         Used when a variable appears in an optional group which is not parsed.
+        Returns a string on error.
```

Comment:
why not raise error on error?

Reply:
The errors are handled by the `Parser` object. This is one of the reasons I think this needs a bit more work.

Reply:
I think the `set_type`/`get_type` interface may be going a bit too far, as it makes it difficult to throw an error in the correct place. The original reason I wanted it was so that the `functional_type` directive knows how many types need to be printed so that it knows whether to put brackets round.

Perhaps keeping `get_type` but moving the parsing into each `TypeableDirective` is the correct interface.

Reply:
I still think that errors are probably the way to go here, let's have a go at this after this stack of PRs is in?

Reply:
This diff is outdated. It now throws errors instead of returning a string. The preceding PR to this implements the suggestion I made in the comment above.